### PR TITLE
Preserving order in Map and Set using doubly linked lists

### PR DESCRIPTION
**Pull request by Perryvw**
Code complicates a bit with the extra list bookkeeping. This does preserve O(1) insert and deletes. Iteration is still O(n), although it does include some extra table lookups, so probably more like O(2n) compared to the O(n) it was before.

### CODE REVIEW

File: src/lualib/Set.ts [hunk 1]
```diff
@@ -6,6 +6,12 @@ Set = class Set<T> {
     private items: { [key: string]: boolean } = {};
```
Comment (tomblind) [inline src/lualib/Set.ts]: Does `Set` even need `items` anymore? `has` could just check `nextKey` or `previousKey`, couldn't it?

Comment (tomblind) [reply src/lualib/Set.ts]: Wait nevermind... I just realized that won't work for a one-item set.

File: test/unit/builtins/map.spec.ts [hunk 2]
```diff
@@ -135,3 +135,45 @@ test("map size", () => {
     expect(util.transpileAndExecute(`let m = new Map([[1,2],[3,4]]); m.clear(); return m.size;`)).toBe(0);
     expect(util.transpileAndExecute(`let m = new Map([[1,2],[3,4]]); m.delete(3); return m.size;`)).toBe(1);
 });
+
+const testMapConstructionCode = `
```
Comment (ark120202) [inline test/unit/builtins/map.spec.ts]: Have you considered using .each?

```ts
test.each(["keys", "values", "entries"])("map.%s() preserves insertion order", (method) => {
    util.testFunction`
        const mymap = new Map();

        mymap.set("x", 1);
        mymap.set("a", 2);
        mymap.set(4, 3);
        mymap.set("b", 6);
        mymap.set(1, 4);
        mymap.set("a", 5);
        
        mymap.delete("b");

        return [...mymap.${method}()];
    `.expectToMatchJsResult();
});
```

Comment (Perryvw) [reply test/unit/builtins/map.spec.ts]: That would mean I can no longer also check for the actual expected value. I have not made my mind up on if that should be there or not yet.

File: src/lualib/Set.ts (hunk 1 above)
Comment (Perryvw) [reply src/lualib/Set.ts]: You do have a point, I think checking nextKey and lastKey would be enough to get rid of items and would reduce storage space required.

Comment (Perryvw) [reply src/lualib/Set.ts]: I made this change, seems fine.

**Pull request merged**